### PR TITLE
CLI kill waits for routers to agree with timeout query parameter

### PR DESCRIPTION
**Pull request by kevo1ution**
## Changes proposed in this PR

- for `waiter kill service-id --timeout TIMEOUT` command: use the timeout query parameter instead of the client side polling method

Here is timing evidence where server sleeps and client is forced to timeout:
```
$ time waiter kill service-id --timeout 5
Killing service w9091-499285cd09532196626361355fe0699d in WAITER1...
Encountered error while killing w9091-499285cd09532196626361355fe0699d in WAITER1.
waiter kill service-id --timeout 5  
0.18s user 
0.02s system 3% cpu 
5.204 total
```

## Why are we making these changes?

- removes the need to poll on the client side and ensures that the command returns when all routers agree that a service has been deleted

depends on #1203

### CODE REVIEW

**Review by shamsimam — COMMENTED** [review]
I'm wondering whether we should continue to have the timeout on the CLI in case the backend server misbehaves.

File: cli/waiter/http_util.py [hunk 1]
```diff
@@ -66,10 +66,13 @@ def __get(url, params=None, read_timeout=None, **kwargs):
     return session.get(url, params=params, timeout=get_timeouts, **kwargs)
 
 
-def __delete(url, params=None, headers=None):
+def __delete(url, params=None, headers=None, read_timeout=None):
     """Sends a DELETE with params to the given url"""
     logging.debug(f'DELETE {url} with params {params} and headers {headers}')
-    return session.delete(url, params=params, timeout=timeouts, headers=headers)
+    get_timeouts = timeouts
```
Comment (shamsimam) [inline cli/waiter/http_util.py:72]: Let's name this `delete_timeouts`

File: cli/waiter/subcommands/kill.py [hunk 2]
```diff
@@ -5,29 +5,29 @@
...
-            if killed:
+            if resp.json().get('routers-agree', False):
                 print(f'Successfully killed {service_id} in {cluster_name}.')
                 return True
             else:
-                print_error('Timeout waiting for service to die.')
-                return False
+                print(f'Successfully killed {service_id} in {cluster_name}, but routers may not be updated yet!')
+                return True
         else:
```
Comment (shamsimam) [inline cli/waiter/subcommands/kill.py:31]: We should differentiate between a failed delete (non-200 response) reported by the server and us timing out and output different error messages.

File: cli/waiter/subcommands/kill.py [hunk 3]
```diff
@@ -122,8 +122,11 @@ def register(add_parser):
...
     parser.add_argument('--force', '-f', help='kill all services, never prompt', dest='force', action='store_true')
-    parser.add_argument('--timeout', '-t', help='timeout (in seconds) for kill to complete',
-                        type=check_positive, default=30)
     parser.add_argument('--service-id', '-s', help='kill by service id instead of token',
                         dest='is-service-id', action='store_true')
+    timeout_group = parser.add_mutually_exclusive_group()
+    timeout_group.add_argument('--timeout', '-t', help='timeout (in seconds) for kill to complete',
+                        type=check_positive, default=30)
+    timeout_group.add_argument('--no-wait', help="does not wait for all routers to confirm deletion", dest='no-wait',
+                        action='store_true')
```
Comment (shamsimam) [inline cli/waiter/subcommands/kill.py:131]: Can this flag be configured to set value of timeout to `0` when configured directly instead of us handling it manually?

File: cli/waiter/subcommands/kill.py [hunk 4]
```diff
@@ -1,36 +1,44 @@
...
-    def service_is_killed():
-        return get_service(cluster, service_id)['status'] == 'Inactive'
-
     cluster_name = cluster['name']
+    http_util.set_retries(0)
     try:
         print(f'Killing service {terminal.bold(service_id)} in {terminal.bold(cluster_name)}...')
-        resp = http_util.delete(cluster, f'/apps/{service_id}')
+        params = {'timeout': timeout_seconds * 1000}
+        read_timeout = 5 if timeout_seconds == 0 else timeout_seconds
```
Comment (shamsimam) [inline cli/waiter/subcommands/kill.py:20]: Why default to 5 if user chose 0 seconds? Avoid passing it (i.e. read_timeout=None) and rely on default global timeout values in that situation?

File: cli/waiter/subcommands/kill.py [hunk 5]
```diff
@@ -1,36 +1,44 @@
...
-                print_error('Timeout waiting for service to die.')
-                return False
+                print(f'Successfully killed {service_id} in {cluster_name}. '
+                      f'Server-side timeout waiting for routers to update.')
+                return True
         else:
             print_error(response_message(resp.json()))
             return False
+    except requests.exceptions.ReadTimeout:
+        message = f'Encountered request read timeout while killing {service_id} in {cluster_name}.'
```
Comment (shamsimam) [inline cli/waiter/subcommands/kill.py:39]: Maybe rephrase to: 
```suggestion
        message = f'Request timed out while killing {service_id} in {cluster_name}.'
```

File: cli/waiter/subcommands/kill.py (hunk 4 above)
Comment (kevo1ution) [reply cli/waiter/subcommands/kill.py:20]: I saw this default being used in ping.py and wasn't sure
https://github.com/twosigma/waiter/blob/02470066f5037bec706b2258adaa046f30e9c6a2/cli/waiter/subcommands/ping.py#L16

File: cli/waiter/subcommands/kill.py [hunk 6]
```diff
@@ -1,36 +1,40 @@
...
+            routers_agree = resp.json().get('routers-agree')
+            if routers_agree:
                 print(f'Successfully killed {service_id} in {cluster_name}.')
                 return True
             else:
-                print_error('Timeout waiting for service to die.')
-                return False
+                print(f'Successfully killed {service_id} in {cluster_name}. '
+                      f'Server-side timeout waiting for routers to update.')
+                return True
```
Comment (shamsimam) [inline cli/waiter/subcommands/kill.py:30]: This should return false since the operation timed-out.

**Review by shamsimam — APPROVED** [review]
